File: _C_satellite/python_old/satelliteAnimation.py
import matplotlib.pyplot as plt
import matplotlib.patches as mpatches
import numpy as np 
import satelliteParam as P


class satelliteAnimation:
    '''
        Create satellite animation
    '''
    def __init__(self):
        self.flagInit = True                  # Used to indicate initialization
        self.fig, self.ax = plt.subplots()    # Initializes a figure and axes object
        self.handle = []                      # Initializes a list object that will
                                              # be used to contain handles to the
                                              # patches and line objects.
        plt.axis([-2.0*P.length, 2.0*P.length, -2.0*P.length, 2.0*P.length])
        plt.plot([-2.0*P.length, 2.0*P.length], [0, 0], 'b--')
        self.length = P.length
        self.width = P.width

    def drawSatellite(self, u):
        # Process inputs to function
        theta = u[0]   # Angle of base, rad
        phi = u[1]     # angle of panel, rad

        self.drawBase(theta)
        self.drawPanel(phi)
#        self.ax.axis('equal') # This will cause the image to not distort

        # After each function has been called, initialization is over.
        if self.flagInit == True:
            self.flagInit = False

    def drawBase(self, theta):
        # points that define the base
        pts =np.matrix([
            [self.width/2.0, -self.width/2.0],
            [self.width/2.0, -self.width/6.0],
            [self.width/2.0 + self.width/6.0, -self.width/6.0],
            [self.width/2.0 + self.width/6.0, self.width/6.0],
            [self.width/2.0, self.width/6.0],
            [self.width/2.0, self.width/2.0],
            [-self.width/2.0, self.width/2.0],
            [-self.width/2.0, self.width/6.0],
            [-self.width/2.0 - self.width/6.0, self.width/6.0],
            [- self.width/2.0 - self.width/6.0, -self.width/6.0],
            [- self.width/2.0, -self.width/6.0],
            [-self.width/2.0, -self.width/2.0]]).T
        R = np.matrix([[np.cos(theta), np.sin(theta)],
                       [-np.sin(theta), np.cos(theta)]])
        pts = R*pts
        xy = np.array(pts.T)

        # When the class is initialized, a polygon patch object will be
        # created and added to the axes. After initialization, the polygon
        # patch object will only be updated.
        if self.flagInit == True:
            # Create the Rectangle patch and append its handle
            # to the handle list
            self.handle.append(mpatches.Polygon(xy, facecolor='blue', edgecolor='black'))
            self.ax.add_patch(self.handle[0]) # Add the patch to the axes
        else:
            self.handle[0].set_xy(xy)         # Update polygon

    def drawPanel(self, phi):
            # points that define the base
            pts = np.matrix([
                [-self.length, -self.width/6.0],
                [self.length, -self.width/6.0],
                [self.length, self.width/6.0],
                [-self.length, self.width/6.0]]).T
            R = np.matrix([[np.cos(phi), np.sin(phi)],
                           [-np.sin(phi), np.cos(phi)]])
            pts = R * pts
            xy = np.array(pts.T)

            # When the class is initialized, a polygon patch object will be
            # created and added to the axes. After initialization, the polygon
            # patch object will only be updated.
            if self.flagInit == True:
                # Create the Rectangle patch and append its handle
                # to the handle list
                self.handle.append(mpatches.Polygon(xy, facecolor='green', edgecolor='black'))
                self.ax.add_patch(self.handle[1])  # Add the patch to the axes
            else:
                self.handle[1].set_xy(xy)  # Update polygon


# Used see the animation from the command line
if __name__ == "__main__":

    simAnimation = satelliteAnimation()    # Create Animate object
    theta = 0.0*np.pi/180                 # Angle of base rad
    phi = 0.0*np.pi/180                   # angle of panel, rad
    simAnimation.drawSatellite([theta, phi, 0, 0])
    #plt.show()
    # Keeps the program from closing until the user presses a button.
    print('Press key to close')
    plt.waitforbuttonpress()
    plt.close()

File: _C_satellite/matlab/hw_a/satelliteAnimation.m
classdef satelliteAnimation
    %
    %    Create satellite animation
    %
    %--------------------------------
    properties
        base_handle
        panel_handle
        length
        width
    end
    %--------------------------------
    methods
        %------constructor-----------
        function self = satelliteAnimation(P)
            self.length = P.length;
            self.width = P.width;
            
            figure(1), clf
            plot([-2*P.length,2*P.length],[0,0],'k--'); % plot track
            hold on
            % initialize the base and panel to initial conditions
            self=self.drawBase(P.theta0);
            self=self.drawPanel(P.phi0);
            axis([-2*P.length, 2*P.length, -2*P.length, 2*P.length]);
        end
        %---------------------------
        function self=update(self, x)
            % Draw satellite is the main function that will call the functions:
            % drawBase and drawPanel create the animation.
            % x is the system state
            theta = x(1);    % angle of base
            phi   = x(2);    % angle of panel

            self=self.drawBase(theta);
            self=self.drawPanel(phi);
            drawnow
        end
        %---------------------------
        function self=drawBase(self, theta)
            % define points on base (without rotation)
            pts = [...
                self.width/2, -self.width/2;...
                self.width/2, -self.width/6;...
                self.width/2+self.width/6, -self.width/6;...
                self.width/2+self.width/6,  self.width/6;...
                self.width/2, self.width/6;...
                self.width/2, self.width/2;...
                -self.width/2, self.width/2;...
                -self.width/2, self.width/6;...
                -self.width/2-self.width/6, self.width/6;...
                -self.width/2-self.width/6, -self.width/6;...
                -self.width/2, -self.width/6;...
                -self.width/2, -self.width/2;...
                ]';
            % define rotation matrix
            R = [cos(theta), sin(theta); -sin(theta), cos(theta)];
            % rotate points
            pts = R*pts;

            if isempty(self.base_handle)
                self.base_handle = fill(pts(1,:),pts(2,:),'b');
            else
                set(self.base_handle,'XData',pts(1,:));
                set(self.base_handle,'YData',pts(2,:));
            end
        end
        %---------------------------
        function self=drawPanel(self, phi)
            % define points on base (without rotation)
            pts = [...
                -self.length, -self.width/6;...
                self.length, -self.width/6;...
                self.length,  self.width/6;...
                -self.length,  self.width/6;...
                ]';
            % define rotation matrix
            R = [cos(phi), sin(phi); -sin(phi), cos(phi)];
            % rotate points
            pts = R*pts;
            
            if isempty(self.panel_handle)
                self.panel_handle = fill(pts(1,:),pts(2,:),'g');
            else
                set(self.panel_handle,'XData',pts(1,:));
                set(self.panel_handle,'YData',pts(2,:));
            end
        end
    end
end